lib/spack/spack/config.py

```diff
@@ -774,13 +821,84 @@ def override(
 
 def _add_platform_scope(cfg: Configuration, name: str, path: str, writable: bool = True) -> None:
     """Add a platform-specific subdirectory for the current platform."""
+    import spack.platforms  # circular dependency
+
     platform = spack.platforms.host().name
     scope = DirectoryConfigScope(
         f"{name}/{platform}", os.path.join(path, platform), writable=writable
     )
     cfg.push_scope(scope)
 
 
+#: Class for the relevance of an optional path conditioned on either
+#: a spec and or explicitly specified as optional.
+class IncludePath(NamedTuple):
+    path: str
+    when: str
+    optional: bool
+
+
+def included_path(entry: Union[str, dict]) -> IncludePath:
+    """Convert the included path entry into an IncludePath.
+
+    Args:
+        entry: include configuration entry
+
+    Returns: converted entry, where an empty ``when`` means the path is
+        not conditionally included
+    """
+    if isinstance(entry, str):
+        return IncludePath(path=entry, when="", optional=False)
+
+    path = entry["path"]
+    when = entry.get("when", "")
+    optional = entry.get("optional", False)
+    return IncludePath(path=path, when=when, optional=optional)
+
+
+def include_path_scope(include: IncludePath) -> Optional[ConfigScope]:
+    """Instantiate an appropriate configuration scope for the given path.
+
+    Args:
+        include: optional include path
+
+    Returns: configuration scope
+
+    Raises:
+        ValueError: included path has an unsupported URL scheme, is required
+            but does not exist; configuration stage directory argument is missing
+        ConfigFileError: unable to access remote configuration file(s)
+    """
+    # circular dependencies
+    import spack.spec
+    import spack.util.path
+    import spack.util.remote_file_cache as rfc_util
+
+    if (not include.when) or spack.spec.eval_conditional(include.when):
+        # canonicalize_path does variable expansion and resolves relative paths
+        include_path = spack.util.path.canonicalize_path(include.path)
+        config_path = rfc_util.local_path(include_path, _include_cache_location)
```

Comment:
Use of `_include_cache_location` is okay for global includes but is not what I thought we'd agreed to for environment includes.